components/locale_canonicalizer/src/locale_canonicalizer.rs

```diff
@@ -579,16 +579,19 @@ impl<'data> LocaleCanonicalizer<'data> {
 
         let mut max = langid.clone();
         self.maximize(&mut max);
+        let variants = max.variants.clone();
         max.variants.clear();
         let mut trial = max.clone();
 
         trial.script = None;
         trial.region = None;
         self.maximize(&mut trial);
         if trial == max {
-            if langid.script.is_some() || langid.script.is_some() {
+            if langid.language != max.language || langid.script.is_some() || langid.region.is_some() {
+                langid.language = max.language;
                 langid.script = None;
                 langid.region = None;
+                langid.variants = variants;
                 return CanonicalizationResult::Modified;
```

Comment:
`nit` would it be faster to separate out the case you're addressing as:

```rust
if langid.language != max.language {
  langid.language = max.language
}
if langid.script.is_some() || langid.region.is_some() {
  langid.script = None;
  langid.region = None;
}
langid.variants = variants;
return CanonicalizationResult::Modified;
```

(it's a rough sketch, so please, rewrite to proper algo logic - my intent is to show that since in most cases we don't operate with divergent languages, we don't need to override them. Not sure about variants)

Reply:
That makes sense! But if I implement it this way, it needs to compare fields 'language', 'script' and 'region' twice and code is getting verbose. Sth like this:
```
if langid.language != max.language || langid.script.is_some() || langid.region.is_some() {
  if langid.language != max.language {
    langid.language = max.language
  }
  if langid.script.is_some() || langid.region.is_some() {
    langid.script = None;
    langid.region = None;
  }
  langid.variants = variants;
  return CanonicalizationResult::Modified;
} else {
  return CanonicalizationResult::Unmodified;
```

And if we check equality first, it still has the same issue. So what do you think about it?